### PR TITLE
SNOW-728803: Do not replace sqlText if it is specified by the user

### PR DESCRIPTION
**Pull request by sfc-gh-ext-simba-lf**
### Description
Regarding issue 137

The PR fixes a special case for submitting queries with requestId where:
1. User executes a request and the requestId is generated
2. The request fails due to a network error
3. Resubmitting the request with the requestId fails because the original request failed in step 2 and sqlText is replaced by the error message

With the PR, the sqlText will not be replaced if the user specifies a sqlText
So even if the original request fails, the connector will still execute the sqlText specified by the user
- Current

```
1. Call INSERT using sqlText that results in a network error
Network error. Could not reach Snowflake.

2. Call INSERT again using sqlText and requestId from step 1
[
  {
    ERRORMESSAGE: 'Error retrieving query results for request id: b15ce10a-88c8-
4cc1-a574-951698da3ce2, please use RESULT_SCAN instead'
  }
]

3. Call SELECT
[]
```

- With PR

```
1. Call INSERT using sqlText that results in a network error
Network error. Could not reach Snowflake.

2. Call INSERT again using sqlText and requestId from step 1
[ { 'number of rows inserted': 1 } ]

3. Call SELECT
[ { C1: 'hello' } ]
```

### Checklist
- [x] Format code according to the existing code style (run `npm run lint:check -- CHANGED_FILES` and fix problems in changed code)
- [x] Create tests which fail without the change (if possible)
- [x] Make all tests (unit and integration) pass (`npm run test:unit` and `npm run test:integration`)
- [x] Extend the README / documentation and ensure is properly displayed (if necessary)
- [x] Provide JIRA issue id (if possible) or GitHub issue id in commit message

### CODE REVIEW

**Comment by sfc-gh-ext-simba-lf** [comment]
Added test for special case when resubmitting a request:
1. Send an initial request that fails with an error
2. Send another request with the requestId and the sqlText from step 1 and passes if the sqlText is not replaced

If the sqlText is replaced, the response from `mock_http_client.js` is an error and the test will fail

File: lib/connection/statement.js [hunk 1]
```diff
@@ -420,6 +420,12 @@ function createContextPreExec(
     {
       statementContext.requestId = statementOptions.requestId;
     }
+
+    // SNOW-728803: In QA mode, set resubmitRequest to true to test that sqlText
+    // is not overwritten if a sqlText is already specified by the user
+    if (statementOptions.requestId === 'SNOW-728803-requestId') {
```
Comment (sfc-gh-ext-simba-lf) [inline lib/connection/statement.js:426]: Only happens in QA mode: https://github.com/snowflakedb/snowflake-connector-nodejs/pull/639/commits/2f6a4abfe6089f78aa8ede769c3159af0ff2a635#diff-9fcac555294d59dbbfef783d617b2d0601df65ca8f061b2bfb6805d4e58c0edfR414

File: test/unit/mock/mock_http_client.js [hunk 2]
```diff
@@ -700,6 +700,114 @@ function buildRequestOutputMappings(clientInfo)
             }
         }
     },
+    {
+      request:
```
Comment (sfc-gh-dprzybysz) [inline test/unit/mock/mock_http_client.js:704]: do we use this mock definition in the new test?

Comment (sfc-gh-ext-simba-lf) [reply test/unit/mock/mock_http_client.js:704]: That mock request corresponds to the first request in the test where only the `sqlText` is specified: https://github.com/snowflakedb/snowflake-connector-nodejs/pull/639/files#diff-abf0655fdf782710cc811e2041916a2f5f3384c5747e42898cf9e5afb8ead03eR960

File: test/unit/snowflake_test.js [hunk 3]
```diff
@@ -936,6 +936,71 @@ describe('connection.execute() statement failure', function ()
...
+        function (callback) {
+          // 2nd request with sqlText and requestId specified
+          statement = connection.execute(
+            {
+              sqlText: sqlText,
+              requestId: requestId,
+              complete: function (err, stmt) {
+                // if there's an error, fail the test with the error
+                if (err) {
+                  done(err);
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_test.js]: `done(err)` also fails the test but I'll change it `callback(err)` for consistency

File: test/unit/snowflake_test.js [hunk 4]
```diff
@@ -936,6 +936,71 @@ describe('connection.execute() statement failure', function ()
...
+                }
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_test.js:982]: Added

File: test/unit/snowflake_test.js [hunk 5]
```diff
@@ -936,6 +936,71 @@ describe('connection.execute() statement failure', function ()
...
+
+                callback();
+              }
+            });
+        },
+        function (callback) {
+          // 2nd request with sqlText and requestId specified
+          statement = connection.execute(
+            {
+              sqlText: sqlText,
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_test.js:976]: The `sqlText` is the same as the first request to test that the `sqlText` is not overwritten when executed with the `requestId`

We could add another test case for when the `sqlText` is different from the first request

File: lib/connection/statement.js (hunk 1 above)
Comment (sfc-gh-ext-simba-lf) [reply lib/connection/statement.js:426]: @sfc-gh-dprzybysz Is there an existing mock SF I can modify for this test? If not, I can look into making one for it

@sfc-gh-pmotacki I don't think there is a specific reason so it's definitely possible to do what you mentioned where `resubmitRequest` is set the same way for both QA and non-QA mode

```
  // use request id passed by user
  if (statementOptions.requestId) {
    statementContext.requestId = statementOptions.requestId;
    statementContext.resubmitRequest = true;
  }
  else {
    // use a random uuid for the statement request id
    statementContext.requestId = uuidv4();
  }
```

Does this look ok?

Comment (sfc-gh-ext-simba-lf) [reply lib/connection/statement.js]: I've remove the test code in production

File: test/unit/snowflake_test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (sfc-gh-dprzybysz) [reply test/unit/snowflake_test.js:962]: yes, let's add a second test with such edge case

[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_test.js]: Added test case

File: test/unit/snowflake_test.js [hunk 6]
```diff
@@ -936,6 +936,96 @@ describe('connection.execute() statement failure', function ()
...
+  });
+
+  it('sqlText is overwritten when resubmitting requests', function (done) {
+    let statement;
+
+    async.series(
+      [
+        function (callback) {
+          // request with only requestId specified
+          statement = connection.execute(
```
Comment (sfc-gh-dprzybysz) [inline test/unit/snowflake_test.js:1005]: statement could be const here (but only if we do a single statement in the test as it now, look at my other comments to this test)

File: test/unit/snowflake_test.js [hunk 7]
```diff
@@ -936,6 +936,96 @@ describe('connection.execute() statement failure', function ()
...
+                    'the execute() callback should be invoked with the statement');
+
+                  callback();
+                }
+              }
+            });
+
+          // the sql text and request id should be the same as what was passed
+          // in
+          assert.strictEqual(statement.getSqlText(), sqlText);
```
Comment (sfc-gh-dprzybysz) [inline test/unit/snowflake_test.js:989]: should we put those assertions in the complete function?

File: test/unit/snowflake_test.js [hunk 8]
```diff
@@ -936,6 +936,96 @@ describe('connection.execute() statement failure', function ()
...
+                assert.ok(err, 'there should be an error');
+                assert.strictEqual(stmt, statement,
+                  'the execute() callback should be invoked with the statement');
+
+                callback();
+              }
+            });
+
+          // the request id should be the same as what was passed in
+          assert.strictEqual(statement.getRequestId(), requestId);
```
Comment (sfc-gh-dprzybysz) [inline test/unit/snowflake_test.js:1020]: should we put this assertion in the complete function?

File: test/unit/snowflake_test.js [hunk 9]
```diff
@@ -936,6 +936,96 @@ describe('connection.execute() statement failure', function ()
   });
 });
 
+describe('connection.execute() with requestId', function () {
+  const connection = snowflake.createConnection(connectionOptions);
+  const sqlText = 'select 1;';
+  const requestId = 'SNOW-728803-requestId';
+
+  before(function (done) {
+    async.series(
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_test.js]: Done

File: test/unit/snowflake_test.js [hunk 10]
```diff
@@ -936,6 +936,96 @@ describe('connection.execute() statement failure', function ()
...
+              complete: function (err, stmt) {
+                assert.ok(err, 'there should be an error');
+                assert.strictEqual(stmt, statement,
+                  'the execute() callback should be invoked with the statement');
+
+                callback();
+              }
+            });
+
+          // the request id should be the same as what was passed in
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_test.js]: Ok

File: test/unit/snowflake_test.js [hunk 11]
```diff
@@ -936,6 +936,96 @@ describe('connection.execute() statement failure', function ()
...
+  });
+
+  it('keep original sqlText when resubmitting requests', function (done) {
+    let statement;
+
+    async.series(
+      [
+        function (callback) {
+          // request with sqlText and requestId specified
+          statement = connection.execute(
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_test.js]: Switched to const

File: test/unit/snowflake_test.js (hunk 7 above)
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_test.js]: The sqlText and requestId are set in the statement even if execution is not completed but the assertions can be moved there

File: test/unit/snowflake_test.js (hunk 8 above)
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_test.js]: Done

File: test/unit/snowflake_test.js [hunk 12]
```diff
@@ -936,6 +936,96 @@ describe('connection.execute() statement failure', function ()
...
+  it('sqlText is overwritten when resubmitting requests', function (done) {
+    let statement;
+
+    async.series(
+      [
+        function (callback) {
+          // request with only requestId specified
+          statement = connection.execute(
+            {
+              // intentionally leave sqlText blank to invoke the connector to overwrite the sqlText
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_test.js:997]: There is no dependency between this test and the previous one. This test leaves `sqlText` blank and expects the connector to overwrite with this: https://github.com/snowflakedb/snowflake-connector-nodejs/pull/639/files#diff-9fcac555294d59dbbfef783d617b2d0601df65ca8f061b2bfb6805d4e58c0edfL1319

File: test/unit/snowflake_test.js (hunk 6 above)
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_test.js]: Done, the test only needs a single statement

File: test/unit/snowflake_test.js [hunk 13]
```diff
@@ -936,6 +936,89 @@ describe('connection.execute() statement failure', function ()
...
+            });
+        }
+      ],
+      function (err) {
+        done(err);
+      });
+  });
+
+  it('sqlText is overwritten when resubmitting requests', function (done) {
+    async.series(
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_test.js:991]: Removed

File: test/unit/snowflake_test.js [hunk 14]
```diff
@@ -936,6 +936,89 @@ describe('connection.execute() statement failure', function ()
...
+      assert.ok(!err, 'there should be no error');
+      assert.strictEqual(conn, connection,
+        'the connect() callback should be invoked with the statement');
+
+      done();
+    });
+  });
+
+  it('keep original sqlText when resubmitting requests', function (done) {
+    async.series(
```
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_test.js]: Removed

File: test/unit/snowflake_test.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (sfc-gh-ext-simba-lf) [reply test/unit/snowflake_test.js]: Yes, the sqlText sent to the server is overwritten and the one on the statement is the same

**Review by sfc-gh-dprzybysz — APPROVED** [review]
